### ISSUE TITLE
Suggestion: rename (alias for backwards compatibility) finalize() to writePact()

### ISSUE THREAD

**Issue by bethesque** [pact-foundation/pact-js#119]
I've noticed some uncertainty about what finalize() does. Someone recently was trying to publish the pacts before calling finalize because they didn't know what the method did. I wonder if a more descriptive name would be helpful?

**Comment by lirantal** [comment]
Agree. Had a similar confusion at first.
Updating the README helped but maybe just a better method name is quicker to grasp before hitting the docs.

**Comment by ivan-kolmychek** [comment]
**TL;DR:** Please, If you will make `writePact()` to behave exactly like `finalize()` now does, either keep the current `writePact()` behavior around, maybe under different name, or provide an option to say to `pact-js`/`pact-node` that it shouldn't try to manage the mock service. 

Just stumbled upon the `writePact` and the fact that it's *not* the same as `finalize` practically saved us from significantly rewriting current setup a lot and/or asking for additional options.

Basically, for multiple reasons, we're using `pact-js` to interface with mock server that is set up separately by other tools. Update from 3.x to 4.x broke our flow - each time we run specs they were failing because pact says that it cannot stop the process with PID `-1`.
The absence of PID is logical, as it doesn't start the process; but trying to stop it in that case is not. :)

After of practically a day of trying to fix it and digging in `pact-node` (as that's where the error was coming from) I've realized that I've stupidly skipped checking out the sources here; so I found that `writePact()` doesn't tries to stop the server.

With it, one small change brought setup back to live. I realize that it's not the usual use-case for `pact-js` and `pact-node`, but in our case it makes sense.

Please don't nuke such useful (for us) method. :)

**Comment by mefellows** [comment]
I'd like to keep flexibility in the API as this allows us to build tools like Mocha UIs on top of it.

That being said, we can probably solve your use case in an alternative way @ivan-kolmychek. Check out pact-web, basically same interface but doesn't try to manage a mock server.

**Comment by ivan-kolmychek** [comment]
@mefellows thanks for the tip. I'm not sure it will be what we need. Can you point me to where can I read more about it?

If I understand correctly, [package.json](https://github.com/pact-foundation/pact-js/blob/f5afc028a2894d41a5fe9d89c10f1c8fca9e9546/package.json.web) for `pact-web` says that it's a `pact for browser`. We're not running `pact` in browser.

Anyhow, we've been using this (a bit weird) setup since Nov 8, 2016, so almost a year now, and it works all right given two conditions - that we don't call `setup()` and that whatever we use to write pacts (`finalize()` before, `writePact()` now) doesn't raises if it cannot stop the server, as it's not really there.

I think we may have to change it in future anyway, so I'll lookg into it, but for now the fact that `writePact()` is there saves us a lot of time. :)

**Comment by ivan-kolmychek** [comment]
By the way, I see you have an issue https://github.com/pact-foundation/pact-js/issues/118 opened. The quirks I've mentioned are related to Docker and the solution from that side is also something that I'm investigating, so maybe I can help with that based on our experience.

**Comment by mefellows** [comment]
It's maybe poorly named (we didn't anticipate this us case for non browser scenarios) but pact web does as you've described. There's nothing browser specific in it though. 

Any help re that Docker would be greatly appreciated

**Comment by ivan-kolmychek** [comment]
> we didn't anticipate this us case for non browser scenarios
So there may be some future breakage if we break the assumption that it runs in browser, like in case there may be need to add something browser-specific.

So I think we would rather stick with just `pact-js`.

From #115 and #118 I've got an impression that you're also checking out setups similar to ours.
> @bethesque a pattern I'm seeing is a multi-docker test harness, where the mock-server is running in a separate container. Might be worth us reviewing docs/other library implementations to see how we can simplify this use case.

Let me try to describe what we have and what problems we've had, maybe it can help someone.

Before I start:
- Please keep in mind that it's used in both development and CI
- The setup is not perfect, we're always looking for ways to improve it, but as it was working fibe, there is almost always something with higher priority. :)

So, in terms of services in `docker-compose.yml` we have:
- "development" service (`dev`) based on `node` where `mocha` is being run in. 
- pact mock service (`pact-mock`)
- selenium service (`selenium`), based on official selenium images (the ones with VNC for additional debug capabilities)
- service that serves an app (`app`), it's just an `nginx` serving static

`dev` depends on `selenium`, `app` and `pact-mock`, `selenium` depends on `pact-mock` and `app`.

All of them belong to the same `default` network since not long ago; before that they were just `link`-ed, as we were constrained to v1 of `docker-compose.yml` 'cause of CI; now it's solved so we've moved to the most recent supported version.

The `dev` service is usually being started by `run` (i.e. `docker-compose run --rm dev /bin/bash`) and it brings up dependant services if they're not up yet.
There can be multiple `dev` services up via `run`, but we've avoided launching more than one `mocha` 'cause we have `pact-mock` and `selenium` as basically shared resources.

So, when `mocha` is being run in `dev` it sets up expectations in `pact-mock` and then drives selenium to visit the app in `app` and app does requests to mock service.

The practical (not "purity") problems that we had running pact-mock inside the same container in v1 of docker-compose.yml was that if we link them like `dev` -> `selenium` -> `dev` it would cause a circular dependency. So we had to break it by providing a third-party services - `app` and `pact-mock`.

Now, as we've moved to versions where networks are supported, we are able to just setup `depends_on`, aliases and access any services from any other services in that networks (basically local to same docker-compose.yml by default, as it creates new network for the group of services).

But the next problem is that, given that `dev` services are set up via `run`, they don't have aliases on them. But good thing is, they can be referenced by their
hostnames, which are available inside them as both `hostname` tool or `$HOSTNAME` env var.

So if we're able to pass the hostname to selenium (as part of URL to visit, for instance), we can ditch both `static` and `pact-mock` as separate servies - `mocha` can manage both starting simple server (`Express.static`) and pact mock service.

Given that I found the "`finalize()` -> `writePacts()`" quickfix, I'll delay a bit looking into it, but so far it can be interesting.

Because we have `pact-mock-service` basically as shared resource for this setup, it may be a problem if we want to paralellize specs execution. So if we'll allow each mocha instance
to have it's own mock service, we eliminate that problem - the second shared service is a selenium, which can handle multiple sessions, if there is enough resources.

Hope it helps you and/or someone else. :)

**Comment by mefellows** [comment]
Thanks @ivan-kolmychek for the detail, that gives a lot of context and whilst there are lots of moving parts, in principle this is nothing unusual about what you're trying to do. This will help us with future features and implementation decisions.

> There's nothing browser specific in it though.

Just re-iterating that `pact-web` is _literally_ the `pact` code, with the bits related to the mock server setup/teardown removed. It was named `pact-web` as it was intended for use cases where the mock server couldn't be started (like a browser/Karma/Jasmine/etc. where the FS can't be accessed).

But either way, you've found your way around the `pact` module to avoid those commands.

**Comment by bethesque** [comment]
Back to the original suggestion, I didn't realise that finalize also stopped the mock service. Ignore my suggestion.

**Comment by mefellows** [comment]
OK, I've flipped and flopped on this one and not sure what I should do going forward. My current thinking is to leave it as is.

In the TS branch, I've simplified the documentation to show the basic 4 functions (and hide the 'advanced' ones). Perhaps this is enough?

**Comment by ivan-kolmychek** [comment]
> In the TS branch, I've simplified the documentation to show the basic 4 functions (and hide the 'advanced' ones). Perhaps this is enough?

Just compared the docs, difference is only in `removeInteractions()` - it is missing from the TS branch version. I've just checked, we do use it before each example to make sure we don't have mocks left over from any previous interaction.

I also have to mention that we use it even though we have significantly rewrote our setup to allow `pact-js` to actually manage mock server, as I described before. :)

I do not understand how removing docs for `removeInteractions()` is it related to the `finalize() vs writePact()` question - I would suggest to keep the docs for `removeInteractions()` and add docs for `writePact()`. 

From my personal perspective and experience in using pact-js, the more granular access you provide the more flexibility that allowes.

Is there anything that prevents, for instance, adding method `stopMockServer()` and make `finalize()` just a shortcut for "writePact() then stopMockServer()"?

**Comment by mefellows** [comment]
It's a fair point. There balance in striking (or trying to) is making it easier for beginners as they are often confused by it all, and having more info just makes it harder. Perhaps increasing flexibility as you say would be useful, and we could deal with the complexity with a wiki page for advanced users or framework developers who need granular access.

FWIW in the TS implementation you should have much more access to things like controlling the mock service

**Comment by ivan-kolmychek** [comment]
> There balance in striking (or trying to) is making it easier for beginners as they are often confused by it all, and having more info just makes it harder. Perhaps increasing flexibility as you say would be useful, and we could deal with the complexity with a wiki page for advanced users or framework developers who need granular access.

This sounds pretty good to me. I think you can have "regular"/"high level" functions suggested in README.md with link below to "all functions" page or something like that.
May I just suggest that, instead of wiki, it can be just another markdown page in codebase? It has multiple benefits, like that it is harder to get it out of sync, as it lives near the code base, it also gets pulled with the code, it can be different in the different branches, etc. etc.
